File: improved-analysis/java/service/ProductService.java
package com.improved.test.service;

import com.improved.test.entity.Product;
import com.improved.test.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Optional;

import com.improved.test.enums.ProductStatus;

@Service
@RequiredArgsConstructor
@Transactional
public class ProductService {

    private final ProductRepository repository;

    public List<Product> findAll() {
        return repository.findAll();
    }

    public Optional<Product> findById(Long id) {
        return repository.findById(id);
    }

    public Product save(Product entity) {
        return repository.save(entity);
    }

    public void deleteById(Long id) {
        repository.deleteById(id);
    }

    public Product suspendProduct(Long id) {
        Product entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));
        entity.suspend();
        return repository.save(entity);
    }

    public Product activateProduct(Long id) {
        Product entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("Product not found with id: " + id));
        entity.activate();
        return repository.save(entity);
    }

}

File: typescript-analysis/src/services/ProductService.ts
import { Product } from '../entities/Product';
import { ProductRepository } from '../repositories/ProductRepository';

export class ProductService {
  private repository: ProductRepository;

  constructor() {
    this.repository = new ProductRepository();
  }

  async create(entity: Product): Promise<Product> {
    return await this.repository.save(entity);
  }

  async findById(id: string): Promise<Product | null> {
    return await this.repository.findById(id);
  }

  async findAll(): Promise<Product[]> {
    return await this.repository.findAll();
  }

  async update(id: string, entity: Partial<Product>): Promise<Product | null> {
    const existing = await this.repository.findById(id);
    if (!existing) return null;

    Object.assign(existing, entity);
    return await this.repository.save(existing);
  }

  async delete(id: string): Promise<void> {
    await this.repository.delete(id);
  }
}
